managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -4039,6 +4122,190 @@ public void clearBacklogFailed(ManagedLedgerException exception, Object ctx) {
+                                log.warn("[{}] Failed to delete offloaded ledger:{} due to",
+                                        name, deletableOffloadedLedger, exception);
+                                counter.countDown();
+                                finishedDeletedOffloadedLedgers.add(deletableOffloadedLedger);
+                                failDeletedOffloadedLedgers.add(deletableOffloadedLedger);
+                            }
+                        });
+            } catch (Exception e) {
+                log.warn("[{}] Failed to retrieve offloaded ledger info of {} due to",
+                        name, deletableOffloadedLedger, e);
+                counter.countDown();
+                finishedDeletedOffloadedLedgers.add(deletableOffloadedLedger);
+                failDeletedOffloadedLedgers.add(deletableOffloadedLedger);
+            }
+        }
+
+        try {
+            if (!counter.await(AsyncOperationTimeoutSeconds, TimeUnit.SECONDS)) {
+                for (Long deletableLedger : deletableLedgers) {
+                    if (!finishedDeletedLedgers.contains(deletableLedger)) {
+                        log.warn("[{}] Failed to delete ledger:{} due to operation timeout", name, deletableLedger);
+                        timeoutDeletedLedgers.add(deletableLedger);
+                    }
+                }
+                for (Long deletableOffloadedLedger : deletableOffloadedLedgers) {
+                    if (!finishedDeletedOffloadedLedgers.contains(deletableOffloadedLedger)) {
+                        log.warn("[{}] Failed to delete offloaded ledger:{} due to operation timeout",
+                                name, deletableOffloadedLedger);
+                        timeoutDeletedLedgers.add(deletableOffloadedLedger);
+                    }
+                }
+            }
+
+            // remove markers after deleting ledgers
+            for (Long ledgerId : succeedDeletedLedgers) {
+                final String deletableLedgerMarker = DELETABLE_LEDGER_MARKER_PREFIX + ledgerId;
+                propertiesMap.remove(deletableLedgerMarker);
+            }
+            for (Long ledgerId : succeedDeletedOffloadedLedgers) {
+                final String deletableLedgerMarker = DELETABLE_OFFLOADED_LEDGER_MARKER_PREFIX + ledgerId;
+                propertiesMap.remove(deletableLedgerMarker);
+            }
+
+            // update retry count to track whether the max limit is reached
+            Set<Long> allFailedLedgers = new HashSet<>();
+            allFailedLedgers.addAll(failDeletedLedgers);
+            allFailedLedgers.addAll(failDeletedOffloadedLedgers);
+            allFailedLedgers.addAll(timeoutDeletedLedgers);
+
+            if (allFailedLedgers.isEmpty()) {
+                log.info("[{}] ledgers: {} and offloaded ledgers: {} are deleted successfully.",
+                        name, deletableLedgers, deletableOffloadedLedgers);
```

Comment:
`deletableLedgers` and `deletableOffloadedLedgers` are Set and not override `toString`, So it will print the object address instead of the items.

Reply:
Those collection objects that inherit `java.util.AbstractCollection` have the ability to return a string representation of this collection (`java.util.AbstractCollection#toString`).